docs/validate.ts

```diff
@@ -0,0 +1,90 @@
+const { fromMarkdown } = require('mdast-util-from-markdown');
+const { mdxFromMarkdown } = require('mdast-util-mdx');
+const { mdx } = require('micromark-extension-mdx');
+// const { visit } = require('unist-util-visit');
+
+// Pure function to return an array of all nodes in the tree
+function flattenTree(node) {
+  const nodes = [node];
+  if (node.children) {
+    return nodes.concat(
+      node.children.flatMap(child => flattenTree(child))
+    );
+  }
+  return nodes;
+}
+
+// Pure predicate function to check if node is a TutorialFallback element
+function isTutorialFallbackNode(node) {
+  const nodeType = node.type;
+  const isFlowElement = nodeType === 'mdxJsxFlowElement';
+  const isJsxTextElement = nodeType === 'mdxJsxTextElement';
+  const isNameTutorialFallback = node.name === 'TutorialFallback';
+  
+  return (isFlowElement || isJsxTextElement) && isNameTutorialFallback;
+}
+
+// Pure function to extract filename from node attributes
+function extractFilename(node) {
+  const fileNameAttr = node.attributes?.find(
+    attr => attr.type === 'mdxJsxAttribute' && attr.name === 'filename'
+  );
+  
+  return fileNameAttr?.value || null;
+}
+export function extractArgumentFileNames(mdxContent) {
+  const tree = fromMarkdown(mdxContent, {
+    extensions: [mdx()],
+    mdastExtensions: [mdxFromMarkdown()]
+  });
+
+  return flattenTree(tree)
+    .filter(isTutorialFallbackNode)
+    .map(extractFilename)
+}
+
+function validateResourceAccess(mdxContent: String, validFilenames: Array<String>) {
+  const argFilenames = extractArgumentFileNames(mdxContent);
+  const validResources = argFilenames.map(argFileName =>
+    validFilenames.includes(argFileName)
+  );
+
+  let hasInvalidResource = false;
```

Comment:
Mutation can be removed by using the existential method (every/some in JS):

```ts
const hasInvalidResource = validResources.some(isValid => !isValid)
```

Theoretically speaking, iterating `validResources` twice (the second time for logging) is less performant than performing both operations in the same iteration (like the current, where it logs and computes `hasInvalidResource` in a single iteration), but pragmatically speaking, the performance improvement is negligible, in this codebase, code maintainability (both readability and extendability) is prioritized over trivial performance improvements.